### PR TITLE
Demonstrative code for detecting Esc keypress

### PR DESCRIPTION
**Pull request by saveriomiroddi**
@piotrmurach Hello Piotr! I'd like to get your comment on this approach to detecting the Esc keypress.

The code is meant to be just demonstrative (it breaks lots of UTs, and the option naming is surely inadequate, plus, I have no experience on terminals), however, I wanted to know if this is the strategy you would apply, as this code fixes my requirement (at least, based on a brief tty-prompt test).

Thanks!

### CODE REVIEW

**Comment by piotrmurach** [comment]
Hi Saverio,

Thanks for taking time to look into implementing this. However, I don't agree with the way you wish to approach the problem. Essentially, the fact that escape key doesn't work is a bug not a feature. Your flag suggests that it is a special case that could be enabled. I've seen similar approach few times before, where the suggestion is to insert a flag and produce code if `a very custom special case` then .... or `some other code`. Escape code is not a special case per say. There is no reason to add new way of reading characters, the recursive call consumes any amount of special characters following the `\e` char. What I think needs to happen is that we need to ensure that `getc` doesn't block when reading characters after special `\e`. In other words, if you see `\e` it may imply a special character then read all the buffered characters quickly and don't block for too long.

File: lib/tty/reader.rb [hunk 1]
```diff
@@ -79,6 +79,8 @@ def initialize(**options)
       @history_exclude    = options.fetch(:history_exclude) { exclude_proc }
       @history_duplicates = options.fetch(:history_duplicates) { false }
 
+      @detect_esc = options.fetch(:detect_esc) { true }
```
Comment (piotrmurach) [inline lib/tty/reader.rb]: Don't need special case as `esc` key not working is a bug, not a feature.

File: lib/tty/reader.rb [hunk 2]
```diff
@@ -152,10 +154,16 @@ def read_keypress(options = {})
...
     def get_codes(options = {}, codes = [])
-      opts = { echo: true, raw: false }.merge(options)
+      opts = { echo: true, raw: false, detect_esc: @detect_esc }.merge(options)
       char = console.get_char(opts)
       handle_interrupt if console.keys[char] == :ctrl_c
-      return if char.nil?
+
+      if char.nil?
+        return
+      elsif @detect_esc
```
Comment (piotrmurach) [inline lib/tty/reader.rb]: It is not a special case, and the code that follows should handle this.

File: lib/tty/reader/console.rb [hunk 3]
```diff
@@ -45,7 +45,17 @@ def initialize(input)
       def get_char(options)
         mode.raw(options[:raw]) do
           mode.echo(options[:echo]) do
-            if options[:nonblock]
+            if options[:detect_esc]
+              begin
```
Comment (piotrmurach) [inline lib/tty/reader/console.rb]: No reason to have this code. This method is called recursively both on unix and windows. I appreciate intention behind it but that's exactly what the rest of `get_codes` does.

**Comment by saveriomiroddi** [comment]
Hello!

Thanks for the review. I posted this PR essentially to understand if the `IO#read_nonblock()` was the correct replacement/addition, but based on what you post, it may/should be a `IO#getc()` instead.

I couldn't find a way to make `getc()` work in the intended way (which would have been the best solution), but I have no experience with consoles :smile: Since I need a workaround I'm keeping my fork hacked and close this PR, however, I'm super happy if a clean fix comes out at some point :smile: 

Many thanks!

**Comment by piotrmurach** [comment]
I'm currently visiting India and I have limited time. However, I wanted to quickly review what you have done to help direct 'the research'. I would definitely forget about the special case flag. I believe, your thinking around nonblocking read is a step in the right direction but probably not in the current form. We're exploring solutions here. 

My thinking goes, when reading characters use `getc` but when reading special characters starting with `\e`, read the next character in non-blocking mode via the recursive call. Could this work? I mean, when the next character is read make sure it is consumed without waiting for the input. There is already `:nonblock` option, maybe use that? Do you have time to investigate?

**Comment by saveriomiroddi** [comment]
Ah great! I'll collect my thoughts and write you over the next couple of days.

I actually spent some time, but I know little, so it's been challenging. I'll make a summary and see how accessing terminal characters be approached (hopefully) differently!

**Pull request closed without merging**